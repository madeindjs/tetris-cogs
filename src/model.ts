export type Point = [x: number, y: number];

export type Line = [from: Point, to: Point];

export type GridSize = [width: number, height: number];

export type Grid = { size: GridSize };

export type ViewBox = [x: number, y: number, width: number, height: number];

export type Link = { points: Line; broken?: boolean };

export enum Rotation {
  Clockwise = 1,
  Anti = -1,
  None = 0,
}

export type Cog = { position: Point; rotation: Rotation };

/**
 * Aggregate of cogs which represent a {@link CogGroupShape}
 */
export type CogGroup = Cog[];

/**
 * Represent all Tetris shapes
 */
export enum CogGroupShape {
  /**
   * ~~~
   * xxxx
   * ~~~
   */
  I = "I",
  /**
   * ~~~
   *  xx
   * xx
   * ~~~
   */
  S = "S",
  /**
   * ~~~
   * x
   * xxx
   * ~~~
   */
  J = "J",
  /**
   * ~~~
   *  x
   * xxx
   * ~~~
   */
  T = "T",
  /**
   * ~~~
   * x
   * x
   * XX
   * ~~~
   */
  L = "L",
  /**
   * ~~~
   * xx
   * xx
   * ~~~
   */
  O = "O",
  /**
   * ~~~
   * xx
   *  xx
   * ~~~
   */
  Z = "Z",
}
