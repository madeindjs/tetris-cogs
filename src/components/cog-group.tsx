import { For, type Accessor } from "solid-js";
import { useLinks } from "../hooks/use-links";
import type { CogGroup } from "../model";
import Cog from "./cog";
import CogsLink from "./cogs-link";

type Props = {
  cogGroup: Accessor<CogGroup>;
  shadow?: boolean;
};

export default function CogGroup({ cogGroup, shadow }: Props) {
  const links = useLinks(cogGroup);
  return (
    <g aria-label="CogGroup">
      <For each={cogGroup()}>
        {(cog) => <Cog position={() => cog.position} size={() => 1} rotation={() => cog.rotation} shadow={shadow} />}
      </For>
      <For each={links()}>{({ points: [from, to], broken }) => <CogsLink from={from} to={to} error={broken} />}</For>
    </g>
  );
}
