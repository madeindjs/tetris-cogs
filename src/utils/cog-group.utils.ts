import { Cog, CogGroup, CogGroupShape, Grid, GridSize, Point } from "../model";
import { getNeighborsCogsBottom, getOppositeRotation } from "./cog.utils";
import { isPointInsideArea, isSamePoint, movePoint } from "./geometry.utils";

function getRandomCogGroupShape(): CogGroupShape {
  const enumValues = Object.values(CogGroupShape);
  const index = Math.floor(Math.random() * enumValues.length);
  return enumValues[index] as CogGroupShape;
}

export function buildCogGroup(origin: Cog, shape: CogGroupShape = getRandomCogGroupShape()): CogGroup {
  const { position, rotation } = origin;

  switch (shape) {
    case CogGroupShape.I:
      return [
        origin,
        { position: movePoint(position, [0, 1]), rotation: getOppositeRotation(rotation) },
        { position: movePoint(position, [0, 2]), rotation },
        { position: movePoint(position, [0, 3]), rotation: getOppositeRotation(rotation) },
      ];
    case CogGroupShape.J:
      return [
        origin,
        { position: movePoint(position, [0, 1]), rotation: getOppositeRotation(rotation) },
        { position: movePoint(position, [1, 1]), rotation },
        { position: movePoint(position, [2, 1]), rotation: getOppositeRotation(rotation) },
      ];
    case CogGroupShape.L:
      return [
        origin,
        { position: movePoint(position, [0, 1]), rotation: getOppositeRotation(rotation) },
        { position: movePoint(position, [0, 2]), rotation },
        { position: movePoint(position, [1, 2]), rotation: getOppositeRotation(rotation) },
      ];
    case CogGroupShape.O:
      return [
        origin,
        { position: movePoint(position, [0, 1]), rotation: getOppositeRotation(rotation) },
        { position: movePoint(position, [1, 0]), rotation: getOppositeRotation(rotation) },
        { position: movePoint(position, [1, 1]), rotation },
      ];
    case CogGroupShape.S:
      return [
        origin,
        { position: movePoint(position, [1, 0]), rotation: getOppositeRotation(rotation) },
        { position: movePoint(position, [0, 1]), rotation: getOppositeRotation(rotation) },
        { position: movePoint(position, [-1, 1]), rotation },
      ];
    case CogGroupShape.T:
      return [
        origin,
        { position: movePoint(position, [-1, 0]), rotation: getOppositeRotation(rotation) },
        { position: movePoint(position, [1, 0]), rotation: getOppositeRotation(rotation) },
        { position: movePoint(position, [0, 1]), rotation: getOppositeRotation(rotation) },
      ];
    case CogGroupShape.Z:
      return [
        origin,
        { position: movePoint(position, [-1, 0]), rotation: getOppositeRotation(rotation) },
        { position: movePoint(position, [0, 1]), rotation: getOppositeRotation(rotation) },
        { position: movePoint(position, [1, 1]), rotation },
      ];
  }
}

export function moveCogGroup(existing: Cog[], group: CogGroup, offset: Point, gridSize: GridSize): CogGroup {
  const newGroup: CogGroup = group.map((c) => ({ ...c, position: movePoint(c.position, offset) }));

  const canMove = newGroup.every((c) => {
    return isPointInsideArea(c.position, gridSize) && !existing.some((e) => isSamePoint(e.position, c.position));
  });

  return canMove ? newGroup : group;
}

function average(numbers: number[]): number {
  let sum = 0;
  for (let i = 0; i < numbers.length; i++) {
    sum += numbers[i];
  }
  return sum / numbers.length;
}

function getCogGroupCenterPoint(group: CogGroup): Point {
  const points = group.map((g) => g.position);
  return [Math.round(average(points.map((p) => p[0]))), Math.round(average(points.map((p) => p[1])))];
}

export function rotateGroup(group: CogGroup) {
  const [cx, cy] = getCogGroupCenterPoint(group);

  let rotatedShape: CogGroup = [];

  for (const cog of group) {
    const x = cog.position[0] - cx;
    const y = cog.position[1] - cy;
    const xNew = cx + y;
    const yNew = cy - x;
    rotatedShape.push({ ...cog, position: [xNew, yNew] });
  }

  return rotatedShape;
}

export function isCogGroupTouchingSomething(cogs: Cog[], group: CogGroup, grid: Grid) {
  if (group.some((c) => c.position[1] === grid.size[1] - 1)) return true;
  return group.some((cog) => getNeighborsCogsBottom(cog, cogs) !== undefined);
}
