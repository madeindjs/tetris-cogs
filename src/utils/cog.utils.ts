import type { Cog, Grid, Point } from "../model";
import { Rotation } from "../model";
import { getNeighborsPoints, isPointInsideArea, isSamePoint, movePoint } from "./geometry.utils";

export function isColliding(cog: Cog, others: Cog[]) {
  return others.some((o) => isSamePoint(o.position, cog.position));
}

export function getNeighborsCogs(cog: Cog, others: Cog[]) {
  const neighborsPoints = getNeighborsPoints(cog.position);
  return others.filter((o) => neighborsPoints.some((p) => isSamePoint(p, o.position)));
}

export function getNeighborsCogsBottom(cog: Cog, others: Cog[]) {
  const point = movePoint(cog.position, [0, 1]);
  return others.find((o) => isSamePoint(point, o.position));
}

export function computeRotationDirection(dir: Rotation, ...dirs: Rotation[]): Rotation {
  return dirs.reduce((_, d) => d * -1, dir);
  for (const d of dirs) dir = d * -1;
  return dir;
}

export function isSameCog(a: Cog, b: Cog) {
  return isSamePoint(a.position, b.position) && a.rotation === b.rotation;
}

export function moveCog(cogs: Cog[], cog: Cog, gridSize: Point, move: Point): Cog {
  const newCog: Cog = { ...cog, position: movePoint(cog.position, move) };
  return !isPointInsideArea(newCog.position, gridSize) || isColliding(newCog, cogs) ? cog : newCog;
}

export function moveCogsToBottom(cogs: Cog[], grid: Grid) {
  return cogs.map((c) => moveCog(cogs, c, grid.size, [0, 1]));
}

export function getOppositeRotation(rotation: Rotation): Rotation {
  return rotation === 0 ? 0 : rotation * -1;
}
