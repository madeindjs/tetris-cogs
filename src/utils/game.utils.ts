import { Rotation, type Cog, type Grid } from "../model";
import { getNeighborsCogs, getOppositeRotation, isSameCog } from "./cog.utils";
import { movePoint } from "./geometry.utils";

export function getCompleteLines(cogs: Cog[], grid: Grid): number[] {
  const points = cogs
    .map((c) => c.position)
    .reduce<Record<number, Set<number>>>((acc, [x, y]) => {
      acc[y] ??= new Set();
      acc[y].add(x);
      return acc;
    }, {});

  return Object.entries(points)
    .filter(([_, ys]) => ys.size === grid.size[0])
    .map(([x]) => Number(x));
}

export function removeLine(cogs: Cog[], y: number) {
  return cogs
    .filter((cog) => cog.position[1] !== y)
    .map<Cog>((cog) => {
      if (cog.position[1] > y) return cog;
      return { ...cog, position: movePoint(cog.position, [0, 1]) };
    });
}

export function checkAndRemoveCompleteLines(cogs: Cog[], grid: Grid) {
  return getCompleteLines(cogs, grid).reduce<Cog[]>((acc, v) => removeLine(acc, v), cogs);
}

export function computeCogsRotation(currentCogs: Cog[]) {
  const cogs = [...currentCogs];
  let changed = false;

  function updateCog(old: Cog, current: Cog) {
    const index = cogs.findIndex((c) => isSameCog(c, old));
    if (index === -1) return console.error("cannot find cog");
    cogs[index] = current;
    changed = true;
  }

  do {
    changed = false;

    for (const cogA of cogs) {
      const others = cogs.filter((c) => !isSameCog(c, cogA));

      for (const cogB of getNeighborsCogs(cogA, others)) {
        if (cogA.rotation === Rotation.None && cogB.rotation !== Rotation.None) {
          updateCog(cogA, { ...cogA, rotation: getOppositeRotation(cogB.rotation) });
        } else if (cogB.rotation === Rotation.None && cogA.rotation !== Rotation.None) {
          updateCog(cogB, { ...cogB, rotation: getOppositeRotation(cogA.rotation) });
        }
      }
    }
  } while (changed);

  return cogs;
}
