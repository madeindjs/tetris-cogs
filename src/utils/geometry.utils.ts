import { Point } from "../model";

export function isCirclesIntersect(x1: number, y1: number, r1: number, x2: number, y2: number, r2: number) {
  // Calculate the distance between the centers of the circles
  const distanceBetweenCenters = Math.sqrt((x2 - x1) ** 2 + (y2 - y1) ** 2);

  // Check if the distance is less than the sum of the radii
  return distanceBetweenCenters < r1 + r2;
}

export function movePoint([x, y]: Point, [tx, ty]: Point): Point {
  return [x + tx, y + ty];
}
