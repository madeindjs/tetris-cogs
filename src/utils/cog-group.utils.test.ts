import { describe, expect, it } from "vitest";
import { Cog, CogGroupShape, Rotation } from "../model";
import { buildCogGroup } from "./cog-group.utils";

describe(buildCogGroup.name, () => {
  const cog: Cog = { position: [0, 0], rotation: Rotation.Clockwise };

  it.each([
    CogGroupShape.I,
    CogGroupShape.J,
    CogGroupShape.L,
    CogGroupShape.O,
    CogGroupShape.S,
    CogGroupShape.T,
    CogGroupShape.Z,
  ])("Should compute shape %s", (shape) => {
    expect(buildCogGroup(cog, shape)).toMatchSnapshot();
  });
});
