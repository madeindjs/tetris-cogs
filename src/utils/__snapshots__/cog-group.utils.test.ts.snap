// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`buildCogGroup > Should compute shape I 1`] = `
[
  {
    "position": [
      0,
      0,
    ],
    "rotation": 1,
  },
  {
    "position": [
      0,
      1,
    ],
    "rotation": -1,
  },
  {
    "position": [
      0,
      2,
    ],
    "rotation": 1,
  },
  {
    "position": [
      0,
      3,
    ],
    "rotation": -1,
  },
]
`;

exports[`buildCogGroup > Should compute shape J 1`] = `
[
  {
    "position": [
      0,
      0,
    ],
    "rotation": 1,
  },
  {
    "position": [
      0,
      1,
    ],
    "rotation": -1,
  },
  {
    "position": [
      1,
      1,
    ],
    "rotation": 1,
  },
  {
    "position": [
      2,
      1,
    ],
    "rotation": -1,
  },
]
`;

exports[`buildCogGroup > Should compute shape L 1`] = `
[
  {
    "position": [
      0,
      0,
    ],
    "rotation": 1,
  },
  {
    "position": [
      0,
      1,
    ],
    "rotation": -1,
  },
  {
    "position": [
      0,
      2,
    ],
    "rotation": 1,
  },
  {
    "position": [
      1,
      2,
    ],
    "rotation": -1,
  },
]
`;

exports[`buildCogGroup > Should compute shape O 1`] = `
[
  {
    "position": [
      0,
      0,
    ],
    "rotation": 1,
  },
  {
    "position": [
      0,
      1,
    ],
    "rotation": -1,
  },
  {
    "position": [
      1,
      0,
    ],
    "rotation": -1,
  },
  {
    "position": [
      1,
      1,
    ],
    "rotation": 1,
  },
]
`;

exports[`buildCogGroup > Should compute shape S 1`] = `
[
  {
    "position": [
      0,
      0,
    ],
    "rotation": 1,
  },
  {
    "position": [
      1,
      0,
    ],
    "rotation": -1,
  },
  {
    "position": [
      0,
      1,
    ],
    "rotation": -1,
  },
  {
    "position": [
      -1,
      1,
    ],
    "rotation": 1,
  },
]
`;

exports[`buildCogGroup > Should compute shape T 1`] = `
[
  {
    "position": [
      0,
      0,
    ],
    "rotation": 1,
  },
  {
    "position": [
      -1,
      0,
    ],
    "rotation": -1,
  },
  {
    "position": [
      1,
      0,
    ],
    "rotation": -1,
  },
  {
    "position": [
      0,
      1,
    ],
    "rotation": -1,
  },
]
`;

exports[`buildCogGroup > Should compute shape Z 1`] = `
[
  {
    "position": [
      0,
      0,
    ],
    "rotation": 1,
  },
  {
    "position": [
      -1,
      0,
    ],
    "rotation": -1,
  },
  {
    "position": [
      0,
      1,
    ],
    "rotation": -1,
  },
  {
    "position": [
      1,
      1,
    ],
    "rotation": 1,
  },
]
`;
